### ISSUE TITLE
REST API functions needed for SWATShare implementation

### ISSUE THREAD

**Issue by dtarb** [hydroshare/hydroshare#299]
- [ ] GET /search/{queryType}/{query}
- [x] GET /scimeta/{pid}
- [x] GET /resource/{pid}
- [x] POST /resource
- [ ] PUT /scimeta/{pid}

**Comment by pkdash** [comment]
We should consider implementing the  API GET /search/{queryType}/{query} only after 'haystack' is integrated into HydroShare.

**Comment by dtarb** [comment]
These were based on a list Venkatesh sent.  I do not know why they want search.  Perhaps it is just to identify the resources of type SWAT that a user has access to.  I'll ask him to clarify and perhaps we can simplify.

**Comment by ironfroggy** [comment]
It would be good to know if the kinds of criteria they require are the same
kind of criteria we'll be allowing filtering by for the Filtering page. If
so, we might have opportunities for code reuse.

On Sat, Apr 11, 2015 at 10:01 PM, David Tarboton <notifications@github.com>
wrote:

> These were based on a list Venkatesh sent. I do not know why they want
> search. Perhaps it is just to identify the resources of type SWAT that a
> user has access to. I'll ask him to clarify and perhaps we can simplify.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hydroshare/hydroshare/issues/299#issuecomment-91968805>
> .
>



-- 


*Calvin Spealman*Lead Front-end Developer

calvin@caktusgroup.com <calvin@caktusgroup.com>
www.caktusgroup.com

**Comment by dtarb** [comment]
@vmmerwade can you please weigh in on this issue

**Comment by lanzh** [comment]
David was right. We would need to query hydroshare and find the SWAT model resources that a user can access using GET.

**Comment by vmmerwade** [comment]
I think the list that we provided was based on the information in the API documentation. If there is a simple way to access the SWAT model instance on hydroshare without implementing search, it should be fine. I am imagining search would be necessary to filter resources at least based on access (public or private).

**Comment by dtarb** [comment]
With this discussion (and I'll include @selimnairb, @horsburgh and @danames  here) we are trying to temper the idealism of the document by the realism of focused and practical use cases.  If you can frame REST API calls (ideally in the form of tests) that SWATShare needs to do its work then we can have coordinated test and purpose driven development around precisely defined intermodule interfaces.  That is really what we are striving for here.  So perhaps an idealized test sequence
* put a swat resource
* Get/search to list swat resources
* Get a swat resource 
can guide the API team.

**Comment by pkdash** [comment]
Using the resource-list api it is possible to get a list of resources that are of specific resource type. The user who is making the request using this api will always get the resources that are viewable by him/her which includes resources created by the user (may be private) any resources shared with the user (public).

**Comment by vmmerwade** [comment]
It seems that, at least to me, the list api should work for what we are trying to do. I will have to talk to Lan to confirm if we need anything more than this. @lanzh  can you comment?

**Comment by lanzh** [comment]
Yes, the resource-list api does what we need.

**Comment by danames** [comment]
I'd like to see this simplified to

GetResource(FilterBy)
PutResource()

Where FilterBy is in the form of a URL Query String like:
"attribute_name=attribute_value&attribute_name=attribute_value"

For example

GetResource("ID=123456")
GetResource("Type=Time_Series&Creator=DanAmes")

**Comment by selimnairb** [comment]
I think we should use different syntax for filtering.  In general, it's not a good idea to allow the HTTP implementation details to leak into the Python client API.  Instead, it should be something like:

    GetResource(ID="123456")
    GetResource(Type="Time_Series", Creator="DanAmes")

We'll let the client API translate between Python syntax and HTTP syntax.

**Comment by selimnairb** [comment]
@lanzh Could you clarify the resource-list API end point you are referring to?

**Comment by lanzh** [comment]
I was referring to the resource-list Pabitra mentioned in this thread. According to what Pabitra described, it can do what we need.

**Comment by pkdash** [comment]
@selimnairb 'resource-list' is the GET /resourceList in the specification.

**Comment by danames** [comment]
I suggested the http querystring syntax because it would allow the
GetResource() function to have a single overload that only requires one
string parameter. This would is extensible in the future to new filter
parameters that may come up. So in other words, shouldn't "ID" and "Type"
qualifiers be part of the string?

GetResource("ID=123456")
GetResource("Type=Time_Series", "Creator=DanAmes")


Of course, since we are talking specifically about the REST API
implementation of GetResource then I suppose it would have to be URL style.
So maybe:

http://hydroshare.org/services/getresource/id=12345
http://hydroshare.org/services/getresource/type=time_series&creator=danames

Anyway, bottom line is that for what we are doing on tools, we need a
simple GetResource REST call that lets us specify any combination of filter
parameters and a PutResource call that lets us add a new resource to
HydroShare...

All of the other API definitions we came up with a year ago are not
necessary as far as I can tell (yet).

- Dan

**Comment by horsburgh** [comment]
@danames - What would the two GetResource() function calls you list return? Seems like the first would return a resource, but what about the second?  What if there are multiple resources of type "time series" with "danames" as the creator?  

Since we are doing all of our access control, cataloging, etc. at the resource level, I think there should be a 1:1 match between a get request and a resource returned.  Resolving multiple permutations of query strings would be difficult, and we don't have a packaging scheme for downloading multiple resources at the same time.  Plus, what if you had access to some but not all of the resources your filter requests? 

We designed separate methods in the API document for getting a resource (getResource) versus getting a list of resources that meet some filter/query criteria (listResources).  The idea was to enable retrieval of a filtered list of resources and then call the GET method for each resource you want to retrieve. I don't think we should try to do both of those things with one overloaded method.

**Comment by danames** [comment]
Looks like the "listResources" method will do what I'm wanting - return a
list of resources that meet some set of filter requirements...

**Comment by selimnairb** [comment]
@yirugi @lanzh @vmmerwade Please update this issue with a list of missing/broken REST API end points needed by SWATShare.  If you can do this by Tues. March 8th, 2016, I can work on this during the hackathon the following day.

**Comment by yirugi** [comment]
We are still waiting for  a REST API to set resource-specific metadata, SWATModel Instance for us. 
I guess we need  PUT /scimeta/{pid} api.

**Comment by selimnairb** [comment]
@yirugi Today @vmmerwade mentioned that resource creation was failing (some metdata fields were not updated).  Could you point me to an example SWAT model that has this problem so that I can test?  Thanks.

**Comment by yirugi** [comment]
@selimnairb  I actually not sure which model he created, but I have a similar problem. On HydroShare resource page, I could only see my models. And also, I tried to create a new SWAT model instance, but after file upload finished, it said 'sorry error occurred~~'.
I tried to access the model, but it also said the error message.

https://www.hydroshare.org/resource/78b6a717c0654cd78fab8cf7e486f425/

**Comment by selimnairb** [comment]
@yirugi I'm not sure about this error.  I would have to see the original file you were uploading.  Can you send me a SWAT model that you can create successfully?  Thanks.

**Comment by mohamedmorsy** [comment]
@selimnairb I tried to created a SWAT Model Instance resource on www.hydroshare.org but I got the following error:

![image](https://cloud.githubusercontent.com/assets/5543781/13580826/e8ad9d70-e471-11e5-816b-ed732d667233.png)


While it works fine on testservices.hydroshare.org. We did not do any new modifications for this resource type. I asked @mjstealey  to take a look on that.

**Comment by yirugi** [comment]
@selimnairb  here's the model I have been used for test
https://www.dropbox.com/s/2r34r4uy0ihor3p/Flat_River.zip?dl=0

Anyway, what @vmmerwade wanted to say was about the REST API setting metadata for SWAT Model Instance resources.

**Comment by selimnairb** [comment]
@yirugi Just so I understand what you are asking for, when you call PUT /scimeta, would you be sending an XML document representing the new science metadata?  This is how the spec is written.

**Comment by yirugi** [comment]
@selimnairb I actually was guessing that the REST API could set metadata for SWAT Model Instance. So, the PUT /scimeta will set SWAT Model Instance metadata? 
Sending XML doc would be fine.

**Comment by selimnairb** [comment]
@yirugi Unfortunately, SWAT Model Instance doesn't extract metadata on creation, so the REST API can't either (the REST API is basically a thin layer on top of various internal APIs).  I will try to get PUT /scimeta done at tomorrow's Hackathon (however there are already competing priorities, so it may not be until next week).

**Comment by yirugi** [comment]
@selimnairb Can you let me know what's the current status?

**Comment by selimnairb** [comment]
@yirugi Sorry, I haven't had a chance to look at this as I have been busy with other projects.  I will try my best to get to it this week.

**Comment by selimnairb** [comment]
@yirugi [This](https://github.com/hydroshare/hydroshare/pull/1103) pull request implements PUT /scimeta for SWAT model instances as well as generic resources.  I have deployed it to playground.hydroshare.org so that you can test it with SWAT share as work continues on this branch.  Let me know how it goes.

**Comment by yirugi** [comment]
@selimnairb I have checked modifying metadata by sending xml file worked. 
I download resourcemetadata.xml file first, add some fields, then upload it via PUT.
One thing I found is the xml file contains a filed 'modified' which has modified time. Should I update the field as well?

**Comment by selimnairb** [comment]
@yirugi No, you don't need to worry about the modification time.  In fact, PUT /scimeta will ignore modification time on update (as it will set the modification time as a consequence of updating the resource metadata).

**Comment by yirugi** [comment]
I have confirmed the PUT /scimeta for SWATModelInstance is working fine on playground site. Can this be deployed to live site?

**Comment by hyi** [comment]
@yirugi Can you test on www and see if this issue is now resolved? If not, what remains to be done?

**Comment by yirugi** [comment]
@hyi I have just tested, but this feature (scimeta) still does not work on www.

**Comment by pkdash** [comment]
@yirugi Are you using the PUT /scimeta REST API?

**Comment by yirugi** [comment]
@pkdash 
Yes, I send a xml file via PUT.

**Comment by pkdash** [comment]
@hyi and @yirugi I am seeing there is a WIP pull request that was created by Brian related to this REST API endpoint. That indicates this API is not ready for use.

**Comment by pkdash** [comment]
@dtarb Should this be closed?

**Comment by yirugi** [comment]
@pkdash I actually have been waiting for this. So, Is the /scimeta for SWATModelInstance working on the live site now?

**Comment by pkdash** [comment]
@yirugi The GET /scimeta/{pid} has been implemented and should work for any resource type. Though the PUT /scimeta/{pid} has been implemented I am not sure if it works for your resource type.

**Comment by yirugi** [comment]
I just confirmed the API worked. I think you can close this issue.

**Comment by pkdash** [comment]
@dtarb As suggested by @yirugi I think you can close it.

**Comment by yirugi** [comment]
Hi, I just found a bug for the creating SWATModelInstance and REST API /scimeta.
After new SWATModelInstance resource is created, if I try to set resource specific metadata using /scimeta API, it fails.

I found that when new SWATModelInstance resource is created, it says :

"Now that your resource has been created the following metadata are still required to publish it or make it public or discoverable:"
- ModelObjective

So, I added anything to the metadata on the webpage. After that, I tried to use /scimeta, then it worked.

I checked resourcemetadata.xml, it didn't have the ModelObjective field. If I add anything, then the field is appeared. I guess it should have empty ModelObjective field, i.e <ModelObjective></ModelObjective>

So, could anyone add the ModelObjective field to SWATModelInstance metadata by default? I think it could solve the problem.

**Comment by pkdash** [comment]
@mohamedmorsy Can you look into it?

**Comment by mohamedmorsy** [comment]
@pkdash Will do

**Comment by jsadler2** [comment]
@pkdash, @yirugi we tested this on the GUI and the functionality for adding a ModelObjective and it showing up in the XML file is working as expectd. 
@yirugi could you clarify exactly what you are doing with more detail? Are you creating the resource via the REST API? Are you updating the metadata using the REST API? If you are using the REST API, could you send us the request content that you are sending?

**Comment by yirugi** [comment]
@jsadler2 Thank you for testing! But I wanted to say that the problem is about missing default ModelObjective tag  right after creating a resource.

To reproduce the error,
1. Create new SWATModelInstance
2. Download its' resource metadata
3. Check if there is <ModelObjective> tag 

For example, I have created a resource :
https://www.hydroshare.org/resource/e7844164b0a3480099513e81f25432e8/
Then download its metadata :
https://www.hydroshare.org/hsapi/scimeta/e7844164b0a3480099513e81f25432e8/

I am trying to create resources, then set resource metadata using REST APIs. I think REST API is fine because after I added the ModelObjective manually, the /scimeta/ PUT API worked correctly.

**Comment by jsadler2** [comment]
@yirugi the modelobjective tag does not appear in the GET /scimeta XML document right after creating the resource, because there is no modelobjective. If the tag were there, the field would be empty. This is true for all resource specific metadata elements, if the field is empty, the tag is not included in the XML. Are you saying that you think the tag should be there even if the field is blank?

**Comment by yirugi** [comment]
@jsadler2 The reason why I though the tag should not be empty is this. if the tag is not available, the /scimeta/ API does not work. If the tag is available (even blank), the API works fine.
If the API works correctly in any circumstance, the tag would not be problem anymore.

**Comment by jsadler2** [comment]
Hi @yirugi, I understand now what you mean. But I cannot replicated the problem. If I make a SWATModelInstance resource and immediately make the /scimeta/ call, it works fine for me even without a model objective. I've attached the file that downloads from the call https://www.hydroshare.org/hsapi/scimeta/af330231d4d240ad949c59efdf9f3ed3/ (this is a resource I made to test this behavior).
[resourcemetadata (3).txt](https://github.com/hydroshare/hydroshare/files/581490/resourcemetadata.3.txt)

**Comment by yirugi** [comment]
Hi @jsadler2 , I have found error parts from my side. Your API works fine.
One small thing is, I wonder if there is an API to change share status from private to public.

**Comment by pkdash** [comment]
@yirugi Yes, there is one for changing the sharing status from private to public. If you are using hs_restclient use the ```setAccessRules()``` function.  If you want use the REST endpoint directly here is the rest endpoint for that which you need to use with a PUT request:

hydroshare.org/hsapi/resource/accessRules/{pid}/

user resource short_id for pid in the above REST endpoint.

**Comment by yirugi** [comment]
@pkdash @jsadler2 Thank you so much for your helps! I'll close this issue.
